### PR TITLE
Add model serialization

### PR DESCRIPTION
**Pull request by ahwillia**
This shouldn't break anything. Adds two improvements:

* Use `m = model.dump()` to get numpy variables for all model variables and `model.load(m)` to reload those variables.
* Puts variable initialization into `model.assign_factors(...)` and `model.assign_warps(...)` so that we don't initialize with zeros only to re-assign the variables with initial values. I thought that I had seen a performance loss in cross-validation, specifically in the model creation step.

Minor:

* Fixes docstring pointed out by @nirum

### CODE REVIEW

File: twpca/model.py [hunk 1]
```diff
@@ -183,10 +161,26 @@ def __init__(self, data, n_components,
         self._recon_cost = tf.reduce_sum(self._mask * (self._pred - self._data)**2) / self.num_datapoints
         self._objective = self._recon_cost + self._regularization
 
-        # initialize values for tensorflow variables
+        # declare which variables are trainable
+        self._trainable_vars = [v for k, v in self._vars.items() if k != 'warp'] # time, neuron, trial factors
```
Comment (poolio) [inline twpca/model.py]: `warp` shouldn't be in `self._vars`, it should only be in `self._params`. After the refactor, everything in `self._vars` should actually be trainable, yes?

File: twpca/model.py [hunk 2]
```diff
@@ -260,12 +259,10 @@ def assign_warps(self, warps):
...
         if warps is not None:
             # warps proved by user
-            if normalize_warps:
-                warps *= self.n_timepoints / np.max(warps)
             shift = warps[:, 0] - 1
             scale = np.ones(self.n_trials)
-            tau = np.hstack((np.ones((self.n_trials, 1)), np.diff(warps, axis=1)))
-            tau = np.maximum(0, tau) # make sure warps are monotonic
+            dt = np.maximum(0, np.diff(warps, axis=1)) # enforce monotonic increasing warps
+            tau = np.hstack((np.ones((self.n_trials, 1)), dt))
```
Comment (poolio) [inline twpca/model.py:267]: what's the point of the np.ones here?

File: twpca/model.py [hunk 3]
```diff
@@ -325,6 +329,10 @@ def fit(self, optimizer=None, niter=1000, lr=1e-3, progressbar=True):
         else:
             raise ValueError("niter and lr must either be numbers or iterables of the same length.")
 
+        # reinitialize all variables if prompted by user
+        if reinitialize:
+            self._sess.run([tf.variables_initializer(v) for v in self._train_var_list])
```
Comment (poolio) [inline twpca/model.py]: I believe you can just do tf.variables_initializer(self._train_ar_list)

File: twpca/model.py [hunk 4]
```diff
@@ -415,6 +423,14 @@ def predict(self, X=None):
 
             return pred
 
+    def dump(self):
+        """Serializes model variables"""
+        return {k : self._sess.run(v) for k, v in self._vars.items()}
+
+    def load(self, new_vars):
+        """Assigns model variables from numpy arrays"""
+        [self._sess.run([tf.assign(self._vars[k], v)]) for k, v in new_vars.items()]
```
Comment (poolio) [inline twpca/model.py]: only call `self._sess.run` once. you could do this by jus having the list comprehension be the argument to `run`.

File: twpca/model.py [hunk 5]
```diff
@@ -415,6 +423,14 @@ def predict(self, X=None):
 
             return pred
 
+    def dump(self):
+        """Serializes model variables"""
+        return {k : self._sess.run(v) for k, v in self._vars.items()}
```
Comment (poolio) [inline twpca/model.py]: you can call `run` on the dict object (`self._vars`) and get back what you want

**Review by poolio — COMMENTED** [review]
Some minor comments. There is typically overhead in performing a `run` call, so you want to avoid calling it in a loop

File: twpca/model.py (hunk 2 above)
Comment (ahwillia) [reply twpca/model.py:267]: `np.diff(...)` gives you one fewer tau than necessary (`shared_length - 1`), so you need to prepend something. If you prepend zeros you are in a much less friendly region of the softplus function (since you pass things backwards through an inverse softplus transform).

File: twpca/utils.py [hunk 6]
```diff
@@ -63,7 +63,11 @@ def stable_rank(matrix):
 
 def inverse_softplus(y):
     """Inverse of the softplus function."""
-    return np.log(np.exp(y + 1e-8) - 1)
+    return np.log(np.exp(y) - 1 + 1e-8)
+
+def softplus(x):
+    """Softplus rectifier function."""
+    np.log(np.exp(x) + 1)
```
Comment (nirum) [inline twpca/utils.py:70]: use `np.log1p(z)` for `np.log(1+z)`, it is more numerically stable

File: twpca/model.py [hunk 7]
```diff
@@ -183,19 +161,36 @@ def __init__(self, data, n_components,
...
 
-        # initialize values for tensorflow variables
+        # initialize optimizer
+        self._lr = tf.placeholder(tf.float32, shape=[])
         self.assign_train_op(optimizer)
-        self.assign_warps(warps)
-        self.assign_factors()
+        utils.initialize_new_vars(self._sess)
 
     def assign_train_op(self, optimizer):
```
Comment (nirum) [inline twpca/model.py:169]: can we use `create_train_op` instead of `assign_train_op`?

File: twpca/model.py [hunk 8]
```diff
@@ -415,6 +425,24 @@ def predict(self, X=None):
 
             return pred
 
+    def dump(self):
+        """Serializes model variables"""
+        return self._sess.run(self._vars)
```
Comment (nirum) [inline twpca/model.py:430]: why not dump to a dictionary? easier for someone to figure out what the variables are, and we don't have to worry about preserving order

File: twpca/model.py [hunk 9]
```diff
@@ -415,6 +425,24 @@ def predict(self, X=None):
...
+
+    def load(self, new_vars):
+        """Assigns model variables from numpy arrays"""
+        self._sess.run([tf.assign(self._vars[k], v) for k, v in new_vars.items()])
```
Comment (nirum) [inline twpca/model.py:434]: You load from a dictionary (`new_vars` is a dictionary like object here), but dump to a list?

**Review by nirum — CHANGES_REQUESTED** [review]

File: twpca/model.py (hunk 8 above)
Comment (ahwillia) [reply twpca/model.py:430]: `self._vars` is a dictionary 😉

Comment (ahwillia) [reply twpca/model.py:430]: I wasn't aware of this but tensorflow nicely returns a dict here.

File: twpca/model.py (hunk 9 above)
Comment (ahwillia) [reply twpca/model.py:434]: See above.

File: twpca/model.py (hunk 7 above)
Comment (ahwillia) [reply twpca/model.py:169]: 👍

File: twpca/utils.py (hunk 6 above)
Comment (ahwillia) [reply twpca/utils.py:70]: nice.

File: twpca/model.py (hunk 8 above)
Comment (nirum) [reply twpca/model.py:430]: ah cool

Comment (nirum) [reply twpca/model.py:430]: What's the difference between `self._vars` and `self._params`?

File: twpca/utils.py (hunk 6 above)
Comment (ahwillia) [reply twpca/utils.py:70]: Is there anything for `np.log(z-1)` out of curiosity?

Comment (nirum) [reply twpca/utils.py:70]: `np.log1p(z-2)`

File: twpca/model.py (hunk 8 above)
Comment (ahwillia) [reply twpca/model.py:430]: `self._vars` are all of the tensorflow variables. The user shouldn't really worry about these, but they are useful to serialize and reinstantiate the model.

`self._params` (which is accessible via `self.params` which gives numpy arrays) holds things the user *does* care about. I.e. the low-dimensional factors and the warping functions.

**Review by nirum — APPROVED** [review]
